### PR TITLE
MINIFICPP-1925 Ensure compatibility with the MiNiFi C2 server

### PR DESCRIPTION
**Pull request by lordgamez**
Ensure that config update and heartbeats from MiNiFi C2 server are handled properly in MiNiFi agent

- Read flowid from configuration update parameter
- Make C2 operation strings case-insensitive
- Handle `null` in requested operations
- Remove `Accept` header temporarily until MiNiFi C2 handling is fixed
- Add test coverage for MiNiFi C2 compatibility

https://issues.apache.org/jira/browse/MINIFICPP-1925

---------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by ihor-sokoliuk-exa** [comment]
Hey @lordgamez 
Great PR!

I've tried your changes, and they do not work with NiFi C2 over HTTPS.
I am getting the next error:

`[2022-10-05 23:07:51.015] [org::apache::nifi::minifi::extensions::curl::HTTPClient] [error] curl_easy_perform() failed SSL peer certificate or SSH remote key was not OK on https://nifi-c2:8880/c2/config/heartbeat, error code 60`

Do you have experience in setting up the secured connection with NiFi C2?

**Comment by lordgamez** [comment]
> Hey @lordgamez Great PR!
> 
> I've tried your changes, and they do not work with NiFi C2 over HTTPS. I am getting the next error:
> 
> `[2022-10-05 23:07:51.015] [org::apache::nifi::minifi::extensions::curl::HTTPClient] [error] curl_easy_perform() failed SSL peer certificate or SSH remote key was not OK on https://nifi-c2:8880/c2/config/heartbeat, error code 60`
> 
> Do you have experience in setting up the secured connection with NiFi C2?

Hi @ihor-sokoliuk-exa ,

I don't have any experience setting up NiFi C2 over SSL, but I'll investigate this use case and try to update the PR with this option.

**Comment by lordgamez** [comment]
Hi @ihor-sokoliuk-exa,

I pushed a new commit 88a3bc2ed9cac02c8d4f0e8caff7ca7b47e01877 that updates the PR to support SSL connection in C2 messages. To enable the SSL connection you need to configure the following:

On C2 server side, in the c2.properties file the `minifi.c2.server.secure=true` should be set and the corresponding keystore and truststore should be configured as well. The `authorizations.yaml` file should also be changed to allow the agent's connections through SSL. For this you should define the agent class's DN in the `authorities.yaml` and set the `allow` action for this class in the `authorizations.yaml` for each endpoint. See the added test files for reference.

On the MiNiFi side you should add an `SSLContextService` controller in the `config.yml` file where you can set the client certificate paths for the SSL connection. After that the `nifi.c2.rest.ssl.context.service` property should be set to the name of the `SSLContextService` in the `minifi.properties` file for the MiNiFi agent to use that SSL service for the C2 connection.

**Comment by ihor-sokoliuk-exa** [comment]
@lordgamez you rock!

Everything works like a charm!

Thank you for your contribution and the instructions particularly.

File: libminifi/src/core/FlowConfiguration.cpp [hunk 1]
```diff
@@ -97,24 +97,24 @@ std::unique_ptr<core::reporting::SiteToSiteProvenanceReportingTask> FlowConfigur
...
     for (auto it = path_split.cbegin(); it != path_split.cend(); ++it) {
       if (*it == "flows" && std::next(it) != path_split.cend()) {
-        flow_id = *++it;
+        payload_flow_id = *++it;
       } else if (*it == "buckets" && std::next(it) != path_split.cend()) {
         bucket_id = *++it;
       }
     }
-    flow_version_->setFlowVersion(url, bucket_id, flow_id);
+    flow_version_->setFlowVersion(url, bucket_id, flow_id ? *flow_id : payload_flow_id);
```
Comment (fgerlits) [inline libminifi/src/core/FlowConfiguration.cpp:117]: Is it normal behavior if the flow ID in the payload is different from the original flow ID?  If not, we could log that something unexpected has happened.

**Review by fgerlits — APPROVED** [review]

File: libminifi/src/core/FlowConfiguration.cpp (hunk 1 above)
Comment (lordgamez) [reply libminifi/src/core/FlowConfiguration.cpp:117]: This is normal in this case when a flow update happens, in that case the C2 server can generate a new flow id for the new flow which is then sent in the payload_flow_id.

File: docker/test/integration/minifi/core/MinifiC2ServerSslContainer.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from .Container import Container
+
+
+class MinifiC2ServerSslContainer(Container):
+    def __init__(self, name, vols, network, image_store, command=None):
+        super().__init__(name, 'minifi-c2-server-ssl', vols, network, image_store, command)
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/minifi/core/MinifiC2ServerSslContainer.py]: Good point, added an `ssl` contructor argument to figure out which version should be instantiated in fd13cbd5ad0819f5a3be0e01cbdb4cac5db3fadd (passing the engine as parameter is more risky as any engine could be passed not just a C2 server engine)

File: docker/test/integration/resources/minifi-c2-server-ssl/config.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+- name: GetFile/success/PutFile
+  id: 098a56ba-f4bf-4323-a3f3-6f8a5e3586bf
+  source id: 2f2a3b47-f5ba-49f6-82b5-bc1c86b96e27
+  source relationship names:
+  - success
+  destination id: e143601d-de4f-44ba-a6ec-d1f97d77ec94
+Controller Services:
+  - name: SSLContextService
+    id: 2438e3c8-015a-1000-79ca-83af40ec1994
+    class: SSLContextService
```
Comment (szaszm) [inline docker/test/integration/resources/minifi-c2-server-ssl/config.yml:33]: This is not referenced anywhere in the file. Is it somehow picked up automatically by the C2 client?

Comment (lordgamez) [reply docker/test/integration/resources/minifi-c2-server-ssl/config.yml:33]: It is retrieved through the `nifi.c2.rest.ssl.context.service` property where its name is set.

**Review by szaszm — APPROVED** [review]

File: docker/test/integration/minifi/core/ImageStore.py [hunk 4]
```diff
@@ -105,6 +107,14 @@ def __build_minifi_cpp_image(self):
                 RUN echo nifi.metrics.publisher.class=PrometheusMetricsPublisher >> {minifi_root}/conf/minifi.properties
                 RUN echo nifi.metrics.publisher.PrometheusMetricsPublisher.port=9936 >> {minifi_root}/conf/minifi.properties
                 RUN echo nifi.metrics.publisher.metrics=RepositoryMetrics,QueueMetrics,GetFileMetrics,GetTCPMetrics,FlowInformation,DeviceInfoNode >> {minifi_root}/conf/minifi.properties
+                RUN echo nifi.c2.enable=true  >> {minifi_root}/conf/minifi.properties
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/minifi/core/ImageStore.py:115]: Added ticket https://issues.apache.org/jira/browse/MINIFICPP-1983

**Review by adamdebreceni — APPROVED** [review]

**Pull request closed without merging**